src/core/inductive.ml

```diff
@@ -0,0 +1,279 @@
+      Pos.none (P_Patt (Some (Pos.none s), ts))
+
+    (** [patt s] creates a pattern without position thanks to the name [s]. *)
+    let patt0: string -> p_term = fun s -> patt s None
+
+    (** [appl t1 t2] creates an application of [t1] to [t2] without
+        position. *)
+    let appl : p_term -> p_term -> p_term = fun t1 t2 ->
+      Pos.none (P_Appl(t1, t2))
+  end
+
+(*
+let create_arg : int -> p_term list = fun len ->
+  let rec aux : int -> p_term list -> p_term list = fun i acc ->
+    if i = 0 then
+      acc
+    else
+      let head =
+        Pos.none (P_Patt (Some (Pos.none ("a" ^ (string_of_int i))), None))
+      in
+      aux (i-1) (head::acc)
+  in
+  aux len []*)
+
+(** [gen_rec_rules ind_sym rec_sym cons_list] returns the p_rules linking
+    with an induction principle [rec_sym] of the inductive type [ind_sym]
+    (with its constructors [cons_list]). *)
+let gen_rec_rules : sym -> sym -> sym list -> p_rule list =
+  fun ind_sym rec_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let app   : p_term -> p_term list -> p_term = List.fold_left P.appl in
+  let fapp  : p_term list -> p_term -> p_term = fun ts f -> app f ts in
+(*  let app2  : sym -> sym list -> p_term = fun t q ->
+    let p_patt = create_patt (Pos.none ("pi" ^ t.sym_name)) in
+    List.fold_left (fun q -> _P_Appl q p_patt)
+  in
+  let fapp2 : p_term -> sym list -> p_term = fun p ts -> app2 p ts  in*)
+
+
+  (* STEP 1: Create the common head of the rules *)
+
+  (*let p_iden = _P_Iden (Pos.none ([], sym_ind.sym_name)) true in
+  let p_patt = create_patt (Pos.none "p")                     in
+  let head =
+    List.map (fun c -> create_patt (Pos.none ("pi" ^ c.sym_name))) cons_list
+  in
+  let common_head = fapp head (_P_Appl p_iden p_patt) in*)
+  let p_iden = P.iden rec_sym.sym_name in
+  let p_patt = P.patt0 "p" in
+  let head   = P.appl p_iden p_patt in
+  let arg : sym list -> p_term = fun l  ->
+    let rec aux : sym list -> p_term -> p_term = fun l acc ->
+      match l with
+      | []   -> acc
+      | t::q ->
+          let p_patt = P.patt0 ("p" ^ t.sym_name) in
+          aux q (P.appl acc p_patt)
+    in
+    aux l head
+  in
+  let common_head = arg cons_list in
+
+  (* STEP 2: Create each p_rule according to a constructor *)
+  (* [gen_rule_cons cons_sym] creates a p_rule according to a constructor
+     [cons_sym]. *)
+  let gen_rule_cons : sym -> p_rule = fun cons_sym ->
+    let p_patt       = P.patt0 ("p" ^ cons_sym.sym_name) in (* RHS *)
+    let t_ident  = P.iden cons_sym.sym_name in (* LHS *)
+    let rec aux : p_term list -> term -> p_term list -> int -> p_rule =
+      fun arg_list t hyp_rec_list i ->
+      match Basics.get_args t with
+      | (Symb(s), ts) ->
+          if s == ind_sym then
+            let (lhs_end, rhs_x_head) =
+              let tmp = fapp (List.rev arg_list) in
+              tmp t_ident, tmp p_patt
+            in
+            let ts = List.map type_to_pattern ts in
```

Comment:
Perhaps we can avoid to inject terms into p_terms by using underscores instead, and rely on the type inference mechanism. Again, you should first specify what gen_rec_rule_cons does in the comment preceding the definition of the function:

With `inductive T:\Pi x1:A1,..,\Pi xm:Am,TYPE := c1:T1 | .. | cn:Tn` we get `rec:\Pi p:(\Pi x1:A1,..,\Pi xm:Am,T x1 .. xm->Prop),U1->..->Un->(\Pi x1:A1,..,\Pi xm:Am,\Pi x:T x1 .. xm, \pi(p x1 .. xm x))` and a rule for each constructor.

If Ti = B1 -> .. -> Bk -> T then the rule for ci is:

rec p p1 .. pn _ .. _ (ci x1 .. xk) --> pi x1 t1? .. xk tk?

with m underscores and tj? = nothing if xj is not recursive
and tj? = (ind_T p p1 .. pn _ .. _ xj) if xj is recursive (Bj = T v1 .. vm)